File: Digital Business Enablement/bluegather/src/main/java/br/com/fiap/bluegather/service/ImagemService.java
package br.com.fiap.bluegather.service;

import br.com.fiap.bluegather.dto.ImagemDTO;
import br.com.fiap.bluegather.model.Evento;
import br.com.fiap.bluegather.model.Imagem;
import br.com.fiap.bluegather.repository.ImagemRepository;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Lazy;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.Objects;
import java.util.Set;

@Service
public class ImagemService {

    @Autowired
    private ImagemRepository imagemRepository;

    @Autowired
    @Lazy
    private EventoService eventoService;

    @Autowired
    private MomentoService momentoService;

    public Page<ImagemDTO> listAll(Pageable pageRequest) {
        Page<Imagem> list = imagemRepository.findAll(pageRequest);
        return list.map(this::convertToDto);
    }

    public ImagemDTO findById(Long id) {
        Imagem entity = findEntityById(id);
        return convertToDto(entity);
    }

    @Transactional
    public ImagemDTO create(ImagemDTO newData) {
        Imagem entity = convertToEntity(newData);
        Imagem savedEntity = imagemRepository.save(entity);
        return convertToDto(savedEntity);
    }

    @Transactional
    public ImagemDTO update(Long id, ImagemDTO updatedData) {
        findEntityById(id);
        updatedData.setId(id);
        Imagem updatedEntity = convertToEntity(updatedData);    
        Imagem savedEntity = imagemRepository.save(updatedEntity);
        return convertToDto(savedEntity);
    }
    
    @Transactional
    public void delete(Long id) {
        Imagem entity = findEntityById(id);
        imagemRepository.delete(entity);
    }

    public Imagem findEntityById(Long id) {
        return imagemRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "(" + getClass().getSimpleName() + ") - Imagem não encontrado(a) por ID: " + id));
    }

    public Set<Imagem> findByEvento(Evento evento) {
        return imagemRepository.findByEvento(evento);
    }
    
    private ImagemDTO convertToDto(Imagem entity) {
        ImagemDTO dto = new ImagemDTO();
        dto.setId(entity.getId());
        dto.setIdEvento(entity.getEvento() != null ? entity.getEvento().getId() : null);
        dto.setIdMomento(entity.getMomento() != null ? entity.getMomento().getId() : null);
        dto.setUrlImagem(entity.getUrlImagem());
        return dto;
    }

    private Imagem convertToEntity(ImagemDTO dto) {
        if (Objects.isNull(dto)) {
            throw new IllegalArgumentException("(" + getClass().getSimpleName() + ") - ImagemDTO não pode ser nulo.");
        }
        Imagem entity;
        if (dto.getId() != null) {
            entity = findEntityById(dto.getId());
        } else {
            entity = new Imagem();
        }
        if (dto.getIdEvento() == null) {
            throw new IllegalArgumentException("(" + getClass().getSimpleName() + ") - ID Evento não pode ser nulo.");
        }
        entity.setEvento(eventoService.findEntityById(dto.getIdEvento()));
        if (dto.getIdMomento() == null) {
            throw new IllegalArgumentException("(" + getClass().getSimpleName() + ") - ID Momento não pode ser nulo.");
        }
        entity.setMomento(momentoService.findEntityById(dto.getIdMomento()));
        entity.setUrlImagem(dto.getUrlImagem());
        return entity;
    }
}

File: Enterprise Application Development/bluegather/Services/ImagemService.cs
using Microsoft.EntityFrameworkCore;
using BlueGather.Dto;
using BlueGather.Models;
using BlueGather.Context;

namespace BlueGather.Services;
public class ImagemService
{
    private readonly BlueGatherContext _context;

    public ImagemService(BlueGatherContext context)
    {
        _context = context;
    }

    public async Task<List<ImagemDto>> FindAll()
    {
        var list = await _context.Imagem
            .Include(x => x.Evento)
            .Include(x => x.Momento)
            .ToListAsync();

        return list.Select(entity => ConvertToDto(entity)).ToList();
    }

    public async Task<ImagemDto> FindById(long id)
    {
        var entity = await FindEntityById(id);
        return ConvertToDto(entity);
    }

    public async Task<ImagemDto> Create(ImagemDto newData)
    {
        var entity = await ConvertToEntity(newData);
        _context.Imagem.Add(entity);
        await _context.SaveChangesAsync();
        return ConvertToDto(entity);
    }

    public async Task<ImagemDto> Update(long id, ImagemDto updatedData)
    {
        var entity = await FindEntityById(id);
        updatedData.Id = entity.Id;
        var updatedEntity = await ConvertToEntity(updatedData);
        _context.Entry(entity).CurrentValues.SetValues(updatedEntity);
        await _context.SaveChangesAsync();
        return ConvertToDto(entity);
    }

    public async Task Delete(long id)
    {
        var entity = await FindEntityById(id);
        _context.Imagem.Remove(entity);
        await _context.SaveChangesAsync();
    }

    public async Task<ImagemModel> FindEntityById(long id)
    {
        var entity = await _context.Imagem
            .Include(x => x.Evento)
            .Include(x => x.Momento)
            .FirstOrDefaultAsync(x => x.Id == id);
        if (entity == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {id}");
        }
        return entity;
    }

    public async Task<List<ImagemDto>> FindByEventoId(long id)
    {
        var list = await _context.Imagem
            .Where(x => x.Evento.Id == id)
            .Include(x => x.Evento)
            .Include(x => x.Momento)
            .ToListAsync();

        return list.Select(x => ConvertToDto(x)).ToList();
    }

    public async Task<List<ImagemDto>> FindByMomentoId(long id)
    {
        var list = await _context.Imagem
            .Where(x => x.Momento.Id == id)
            .Include(x => x.Evento)
            .Include(x => x.Momento)
            .ToListAsync();

        return list.Select(x => ConvertToDto(x)).ToList();
    }

    private ImagemDto ConvertToDto(ImagemModel entity)
    {
        return new ImagemDto
        {
            Id = entity.Id,
            IdEvento = entity.Evento.Id,
            IdMomento = entity.Momento.Id,
            UrlImagem = entity.UrlImagem
        };
    }

    private async Task<ImagemModel> ConvertToEntity(ImagemDto dto)
    {
        var evento = await _context.Evento.FindAsync(dto.IdEvento);
        if (evento == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {dto.IdEvento}");
        }

        var momento = await _context.Momento.FindAsync(dto.IdMomento);
        if (momento == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {dto.IdMomento}");
        }

        return new ImagemModel
        {
            Id = dto.Id ?? 0,
            Evento = evento,
            Momento = momento,
            UrlImagem = dto.UrlImagem
        };
    }
}